worlds/pokemon_crystal/__init__.py

```diff
@@ -0,0 +1,325 @@
+            badge_locs = [loc for loc in self.multiworld.get_locations(self.player) if "Badge" in loc.tags]
+            badge_items = [self.create_item_by_code(loc.default_item_code) for loc in badge_locs]
+
+            # 5/8 badge locations in each region do not require a HM to access, so only trying once should be okay.
+            # I generated 1000 seeds with shuffled badges and none of them broke here, so it's fine probably
+            self.random.shuffle(badge_locs)
+            collection_state = self.multiworld.get_all_state(False)
+            fill_restrictive(self.multiworld, collection_state, badge_locs, badge_items,
+                             single_player_placement=True, lock=True, allow_excluded=True)
+
+    def generate_output(self, output_directory: str) -> None:
+
+        self.generated_pokemon = copy.deepcopy(crystal_data.pokemon)
+        self.generated_starters = (["CYNDAQUIL", "QUILAVA", "TYPHLOSION"],
+                                   ["TOTODILE", "CROCONAW", "FERALIGATR"],
+                                   ["CHIKORITA", "BAYLEEF", "MEGANIUM"])
+        self.generated_starter_helditems = ("BERRY", "BERRY", "BERRY")
+        self.generated_trainers = copy.deepcopy(crystal_data.trainers)
+        self.generated_misc = copy.deepcopy(crystal_data.misc)
+        self.generated_tms = copy.deepcopy(crystal_data.tmhm)
+        self.generated_wild = copy.deepcopy(crystal_data.wild)
+        self.generated_static = copy.deepcopy(crystal_data.static)
+        self.generated_palettes = {}
+        self.generated_phone_traps = []
+        self.generated_phone_indices = []
+        self.generated_music = []
+        self.generated_wooper = "WOOPER"
+        # self.generated_sfx = copy.deepcopy(crystal_data.sfx.pointers)
+
+        randomize_pokemon(self)
+
+        if self.options.randomize_starters.value:
+            randomize_starters(self)
+
+        if self.options.randomize_tm_moves.value:
+            randomize_tms(self)
+
+        if self.options.randomize_trainer_parties.value:
+            randomize_trainers(self)
+        elif self.options.randomize_learnsets.value:
+            vanilla_trainer_movesets(self)
+
+        if self.options.randomize_wilds.value:
+            randomize_wild_pokemon(self)
+
+        if self.options.randomize_static_pokemon.value:
+            randomize_static_pokemon(self)
+
+        if self.options.randomize_music.value:
+            music_pool = [music_id for music_name, music_id in crystal_data.music.consts.items() if
+                          music_name != "MUSIC_NONE"]
+            for _music in crystal_data.music.maps:
+                new_music = self.random.choice(music_pool)
+                self.generated_music.append(new_music)
+
+        # if self.options.randomize_sfx:
+        #     self.random.shuffle(self.generated_sfx)
+
+        if self.options.enable_mischief.value:
+            misc_activities(self)
+
+        generate_phone_traps(self)
```

Comment:
This could be put inside a `if self.options.phone_trap_weight:` like many of the others here, and then the actual function could have it taken out